tktitler.py

```diff
@@ -80,39 +123,121 @@ def postfix(titletupel, gfyear=gfyear, type=POSTFIXTYPE_SINGLE):
+    gfyear = _validate(title, gfyear)
 
-    root, period = titletupel
+    root, period = title
 
     replace_dict = {'æ': 'ae', 'ø': 'oe', 'å': 'aa',
                     'Æ': 'AE', 'Ø': 'OE', 'Å': 'AA'}
     root = _multireplace(root, replace_dict)
 
-    prefix_ = ""
-    postfix_ = ""
+    prefix = ""
+    postfix = ""
     if type == EMAILTYPE_POSTFIX:
-        postfix_ = str(period)[2:4]
+        postfix = str(period)[2:4]
     elif type == EMAILTYPE_PREFIX:
-        prefix_ = prefix(("", period), gfyear, type=PREFIXTYPE_NORMAL)
+        prefix = tk_prefix(("", period), gfyear, type=PREFIXTYPE_NORMAL)
     else:
         raise ValueError("\'%s\' is not a valid type-parameter" % type)
-    return "" + prefix_ + root + postfix_
+    assert isinstance(prefix + root + postfix, str)
+    return prefix + root + postfix
+
+
+def _normalize(input_alias):
+    table = {'$': 'S',
+             '\N{POUND SIGN}': 'S',
+             '\N{DOUBLE-STRUCK CAPITAL C}': 'C'}
+
+    def tr(c):
+        try:
+            return table[c]
+        except KeyError:
+            try:
+                return str(unicodedata.digit(c))
+            except ValueError:
+                return c
+
+    s = input_alias.upper()
+    return re.sub(r'[^0-9A-ZÆØÅ]', lambda mo: tr(mo.group(0)), s)
+
+
+def _parse_prefix(prefix):
+    pattern = r"^([KGBOT][KGBOT0-9]*)?$"
+    if not re.match(pattern, prefix):
+        raise ValueError(prefix)
+    prefix_value = dict(K=-1, G=1, B=2, O=3, T=1)
+    factors = []
+    for base, exponent in re.findall(r"([KGBOT])([0-9]*)", prefix):
+        factors.append(int(exponent or 1) * prefix_value[base])
+    return sum(factors)
 
 
-def parse(alias, gfyear=gfyear):
-    pass  # return (root, period)
+def _parse_postfix(postfix):
+    if not isinstance(postfix, str):
+        raise TypeError(type(postfix))
+    if not postfix:
+        return
+    if len(postfix) == 2:
+        v = int(postfix)
+        return 2000 + v if v < 56 else 1900 + v
+    elif len(postfix) == 4:
+        first, second = int(postfix[0:2]), int(postfix[2:4])
+        # Note that postfix 1920, 2021 and 2122 are technically ambiguous,
+        # but luckily there was no BEST in 1920 and this script hopefully
+        # won't live until the year 2122, so they are not actually
+        # ambiguous.
+        if postfix == '2021':
+            # TODO: Should '2021' be parsed as 2020/21 or 2021/22?
+            raise NotImplementedError(postfix)
+        if (first + 1) % 100 == second:
+            # There should be exactly one year between the two numbers
+            return 2000 + first if first < 56 else 1900 + first
+        elif first in (19, 20):
+            # 19xx or 20xx
+            return int(postfix)
+        else:
+            raise ValueError(postfix)
+    else:
+        # Length is neither 2 nor 4
+        raise ValueError(postfix)
 
 
-def _validate(titletupel, gfyear):
-    root, period = titletupel
+def parse_relative(input_alias):
```

Comment:
Det virker som en naturlig subrutine for `parse`, og det kan blive nyttigt hvis vi f.eks. vil lave en mere intelligent håndtering af EFUIT-titler (hvor det f.eks. kan gælde at GEFUIT14 er den samme som EFUIT12).

Funktionen `parse` er nok den mest anvendelige, men `parse_relative` er en lidt mindre lossy parsing af aliaset. Man kunne forestille sig at lave en lossless parser (så enhver streng der repræsenterer en titel, såsom "K2TOGFUET1516", parses til en struktur der indeholder nok information til at rekonstruere den præcis samme streng), og ud fra denne vil `parse_relative` være en simpel nedkogning af parserens output, og `parse` vil ligeledes (som nu) være en nedkogning af `parse_relative`.